### PR TITLE
Improve C++ interface for tallies, filters, materials

### PR DESCRIPTION
**Pull request by paulromano**
A lot of our C API functions are devoted to making changes to and accessing information for tallies, filters, and materials. By virtue of having C interfaces, many of these functions are not very pleasant to use from C++. This PR formalizes many of the C API functions as proper getters/setters on the underlying C++ classes and significantly improves the usability. Some of the highlights:

- Filters and tallies can be created with `Filter::create()` and `Tally::create()` static methods which return non-owning pointers to a new filter or tally.
- Filters no longer have the awkward `initialize()` method that performs some initialization actions _after_ the object is already constructed.
- The `set_id()` method in each class now handles changes to unique IDs seamlessly, including auto-assignment and updating the associated ID->index map.
- Data members for filters are now private/protected.
- Many of the `fatal_error`s are now exceptions so that user code has the possibility of recovering.

These changes will significantly simplify calls into libopenmc from coupling applications written in C++.

As a side note, I've started making judicious use of `gsl::span` and `gsl::index` in the code here.

### CODE REVIEW

File: src/tallies/tally.cpp [hunk 1]
```diff
@@ -240,37 +240,295 @@ score_str_to_int(std::string score_str)
...
+
+  // Determine number of filters
+  std::vector<int> filter_ids;
+  if (check_for_node(node, "filters")) {
+    filter_ids = get_node_array<int>(node, "filters");
+  }
+
+  // Allocate and store filter user ids
+  std::vector<Filter*> filters;
+  if (!filter_ids.empty()) {
```
Comment (amandalund) [inline src/tallies/tally.cpp]: This outer if statement is redundant.

File: src/tallies/filter_legendre.cpp [hunk 2]
```diff
@@ -10,7 +10,13 @@ namespace openmc {
...
 LegendreFilter::from_xml(pugi::xml_node node)
 {
-  order_ = std::stoi(get_node_value(node, "order"));
+  this->set_order(std::stoi(get_node_value(node, "order")));
+}
+
+void
+LegendreFilter::set_order(int order)
+{
+  order_ = order;
```
Comment (amandalund) [inline src/tallies/filter_legendre.cpp]: Maybe add a check here (and in the other functional expansion filters) to make sure the order is >= 0.

File: src/tallies/tally.cpp (hunk 1 above)
Comment (paulromano) [reply src/tallies/tally.cpp]: Indeed it is. Thanks for pointing that out!

File: src/tallies/filter_sph_harm.cpp [hunk 3]
```diff
@@ -12,21 +12,34 @@ namespace openmc {
...
+void
+SphericalHarmonicsFilter::set_cosine(gsl::cstring_span cosine)
+{
+  if (cosine == "scatter") {
+    cosine_ = SphericalHarmonicsCosine::scatter;
+  } else if (cosine == "particle") {
+    cosine_ = SphericalHarmonicsCosine::particle;
+  } else {
+    std::stringstream err_msg;
+    err_msg << "Unrecognized cosine type, \"" << cos
```
Comment (amandalund) [inline src/tallies/filter_sph_harm.cpp]:
```suggestion
    err_msg << "Unrecognized cosine type, \"" << cosine
```

**Pull request merged**